core/src/cluster_info_entries_listener.rs

```diff
@@ -0,0 +1,174 @@
+};
+use solana_ledger::blockstore::Blockstore;
+use solana_ledger::leader_schedule_utils::slot_leader_at;
+use solana_metrics::inc_new_counter_debug;
+use solana_poh::poh_recorder::PohRecorder;
+use solana_sdk::{clock::Slot, pubkey::Pubkey};
+use std::{
+    cmp::Ordering::{Equal, Greater},
+    collections::{hash_map::Entry, HashMap},
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        {Arc, Mutex},
+    },
+    thread::{self, sleep, Builder, JoinHandle},
+    time::Duration,
+};
+
+pub type DuplicateShredSender = CrossbeamSender<Pubkey>;
+pub type DuplicateShredReceiver = CrossbeamReceiver<Pubkey>;
+
+pub struct ClusterInfoEntriesListener {
+    thread_hdls: Vec<JoinHandle<()>>,
+}
+
+impl ClusterInfoEntriesListener {
+    pub fn new(
+        exit: &Arc<AtomicBool>,
+        cluster_info: Arc<ClusterInfo>,
+        poh_recorder: &Arc<Mutex<PohRecorder>>,
+        blockstore: Arc<Blockstore>,
+        duplicate_slot_sender: DuplicateSlotSender,
+    ) -> Self {
+        let (duplicate_shred_sender, duplicate_shred_receiver) = unbounded();
+        let exit_ = exit.clone();
+        let cluster_info_ = cluster_info.clone();
+
+        let listen_thread = Builder::new()
+            .name("solana-cluster_info_entries_listener".to_string())
+            .spawn(move || {
+                let _ = Self::recv_loop(exit_, &cluster_info_, duplicate_shred_sender);
+            })
+            .unwrap();
+
+        let exit_ = exit.clone();
+        let poh_recorder = poh_recorder.clone();
+        let duplicate_shreds_thread = Builder::new()
+            .name("solana-cluster_info_process_duplicate_shreds".to_string())
+            .spawn(move || {
+                let _ = Self::process_duplicate_shreds_loop(
+                    exit_,
+                    &cluster_info,
+                    duplicate_shred_receiver,
+                    poh_recorder,
+                    blockstore,
+                    duplicate_slot_sender,
+                );
+            })
+            .unwrap();
+
+        Self {
+            thread_hdls: vec![listen_thread, duplicate_shreds_thread],
+        }
+    }
+
+    pub fn join(self) -> thread::Result<()> {
+        for thread_hdl in self.thread_hdls {
+            thread_hdl.join()?;
+        }
+        Ok(())
+    }
+
+    fn recv_loop(
+        exit: Arc<AtomicBool>,
+        cluster_info: &ClusterInfo,
+        duplicate_shred_sender: DuplicateShredSender,
+    ) -> Result<()> {
+        let mut cursor = Cursor::default();
+        while !exit.load(Ordering::Relaxed) {
+            let entries = cluster_info.get_entries(&mut cursor);
+            inc_new_counter_debug!("cluster_info_entries_listener-recv_count", entries.len());
+            if !entries.is_empty() {
+                entries.iter().for_each(|value| {
+                    if let CrdsData::DuplicateShred(_, _) = value.data {
+                        if let Err(e) = duplicate_shred_sender.send(value.pubkey()) {
```

Comment:
any reason not to just directly reconstruct the shred proofs here? i.e. move the work from `process_duplicate_shreds_loop()` here